### PR TITLE
feat(release): add GitLab support

### PR DESCRIPTION
**Pull request by gautamprikshit1**
Adding Support for Gitlab

### CODE REVIEW

**Comment by MarcoIeni** [comment]
I ran `release-plz release --backend gitlab --git-release --git-token $GITLAB_TOKEN` with a Personal Access created [here](https://gitlab.com/-/profile/personal_access_tokens), but I get the following error:

```
HTTP status client error (403 Forbidden) for url (https://gitlab.com/api/v4/projects/MarcoIeni%2Fmarco-test-gitlab/releases)
```

Did you manage to create a release with a personal access token? Which token should I use? Am I doing something wrong? 😅

**Comment by gautamprikshit1** [comment]
I will take another look at it.

**Comment by MarcoIeni** [comment]
by having a first look at the docs, it seems that we need to differentiate among token types.
https://docs.gitlab.com/ee/api/releases/#authentication

Also, it seems that the header should be called `PRIVATE-TOKEN` for personal access tokens. Instead we are using `authorization`.

Can you try to test this in GitLab?

**Comment by MarcoIeni** [comment]
It's working https://gitlab.com/MarcoIeni/marco-test-gitlab/-/releases#v0.1.0
There's just a small error in the header format. 👍

File: crates/release_plz_core/src/gitlab_client.rs [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+    pub fn default_headers(&self) -> anyhow::Result<HeaderMap> {
+        let mut headers = HeaderMap::new();
+        headers.insert(
+            reqwest::header::HeaderName::from_static("content-type"),
+            reqwest::header::HeaderValue::from_static("application/json"),
+        );
+        let auth_header: HeaderValue =
+            format!("private-token: {}", self.remote.token.expose_secret())
+                .parse()
+                .context("Invalid Gitlab token")?;
```
Comment (MarcoIeni) [inline crates/release_plz_core/src/gitlab_client.rs:33]:
```suggestion
        let auth_header: HeaderValue = self
            .remote
            .token
            .expose_secret()
            .parse()
            .context("Invalid Gitlab token")?;
```

File: crates/release_plz_core/src/backend.rs [hunk 2]
```diff
@@ -16,13 +16,23 @@ use tracing::{debug, info, instrument};
 pub enum GitBackend {
     Github(GitHub),
     Gitea(Gitea),
+    Gitlab(GitLab),
+}
+
+#[derive(Serialize)]
+pub struct GitlabReleaseOption<'a> {
+    #[serde(rename = "ref")]
+    ref_field: &'a str,
```
Comment (MarcoIeni) [inline crates/release_plz_core/src/backend.rs:25]:
```suggestion
```
not needed

File: crates/release_plz_core/src/backend.rs [hunk 3]
```diff
@@ -169,6 +180,25 @@ impl GitClient {
             body,
             name: tag,
         };
+
+        if matches!(self.backend, BackendType::Gitlab) {
+            let gitlab_release_options = GitlabReleaseOption {
+                ref_field: "main",
```
Comment (MarcoIeni) [inline crates/release_plz_core/src/backend.rs:186]:
```suggestion
```

**Review by MarcoIeni — APPROVED** [review]
Thanks!

**Comment by MarcoIeni** [comment]
I just did two follow-up PRs:
- https://github.com/MarcoIeni/release-plz/pull/612
- https://github.com/MarcoIeni/release-plz/pull/613

**Pull request merged**